### ISSUE TITLE
tf.estimator.train_and_evaluate got wrong Evaluation accuray and loss

### ISSUE THREAD

**Issue by honeytidy** [tensorflow/tensorflow#21572]
I use tf.estimator.train_and_evaluate to train and evaluate my model. This is my code:

```
def model_fn(features, labels, mode):
    is_training = (mode == tf.estimator.ModeKeys.TRAIN)
    with slim.arg_scope(resnet_v2.resnet_arg_scope()):
        logits, endpoints = resnet_v2.resnet_v2_50(features, num_classes=cls_num,
                is_training=is_training)
    logits = tf.squeeze(logits, [1, 2])
    preds = tf.argmax(logits, 1)
    loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
    accuracy = tf.metrics.accuracy(labels=labels, predictions=preds)
    metrics = {'accuracy': accuracy}
    if mode == tf.estimator.ModeKeys.EVAL:
        return tf.estimator.EstimatorSpec(mode, loss=loss, eval_metric_ops=metrics)
    optimizer = tf.train.AdagradOptimizer(learning_rate=0.1)
    train_op = optimizer.minimize(loss, global_step=tf.train.get_global_step())
    return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op)

def train_input_fn():
    dataset = tf.data.Dataset.from_tensor_slices((np.array(img_train), label_train))
    dataset = dataset.repeat(5).batch(8)
    return dataset

def eval_input_fn():
    dataset = tf.data.Dataset.from_tensor_slices((np.array(img_test), label_test))
    dataset = dataset.repeat(1).batch(8)
    return dataset

estimator = tf.estimator.Estimator(model_fn=model_fn, model_dir='logs')
train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn)
eval_specs = tf.estimator.EvalSpec(input_fn=eval_input_fn)
tf.estimator.train_and_evaluate(estimator, train_spec, eval_specs)
```
The training step is ok and the loss became very small (about 0.001), but the evaluation result is wrong (the following is the evaluaiton log):

```
...
INFO:tensorflow:Saving dict for global step 625: accuracy = 0.5, global_step = 625, loss = 1330830600000.0
...
```
The task is very simple, just a binaray classfication. I don not think it is overfitting. Is there something wrong for my evaluation code?

**Comment by honeytidy** [comment]
@aravind3134 decreasing the learning rate to 0.001 seems not work:
```
INFO:tensorflow:Saving dict for global step 625: accuracy = 0.5, global_step = 625, loss = 7749.7485
```
The loss is smaller but the accuracy is also 0.5.

**Comment by aravind3134** [comment]
Can you let me know the data-set size and can I see more logs in the console, if you have any.

You have them probably above this line:
INFO:tensorflow:Saving dict for global step 625: accuracy = 0.5, global_step = 625, loss = 7749.7485

If you don't find any, you can edit line 
train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn)   to

logging_hook = tf.train.LoggingTensorHook (tensors=tensors_to_log, every_n_iter=50)
train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn, max_steps=1000, hook = [logging_hook])

This should give you the summary of the training you are doing. 

See if your loss during your training is fluctuating. If it is, you might have to look for a better learning rate, may be by performing a grid search on the interval [0.00001, 0.001]. 

The loss seems to have decreased a lot, but nothing can be judged with one line. Let me know more about the logs.

Hope this helps.

**Comment by honeytidy** [comment]
OK, this is my training log:

```
INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_save_checkpoints_secs': 600, '_session_config': None, '_keep_checkpoint_max': 5, '_task_type': 'worker', '_global_id_in_cluster': 0, '_is_chief': True, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7fefaadf1ed0>, '_evaluation_master': '', '_save_checkpoints_steps': None, '_keep_checkpoint_every_n_hours': 10000, '_service': None, '_num_ps_replicas': 0, '_tf_random_seed': None, '_master': '', '_device_fn': None, '_num_worker_replicas': 1, '_task_id': 0, '_log_step_count_steps': 100, '_model_dir': 'logs', '_train_distribute': None, '_save_summary_steps': 100}
INFO:tensorflow:Running training and evaluation locally (non-distributed).
INFO:tensorflow:Start train and evaluate loop. The evaluate will happen after every checkpoint. Checkpoint frequency is determined based on RunConfig arguments: save_checkpoints_steps None or save_checkpoints_secs 600.
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
2018-08-16 09:47:59.546073: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-08-16 09:48:01.441117: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties: 
name: Tesla P40 major: 6 minor: 1 memoryClockRate(GHz): 1.531
pciBusID: 0000:83:00.0
totalMemory: 22.38GiB freeMemory: 22.21GiB
2018-08-16 09:48:01.441186: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0
2018-08-16 09:48:01.811401: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-08-16 09:48:01.811463: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 
2018-08-16 09:48:01.811477: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N 
2018-08-16 09:48:01.812022: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 21544 MB memory) -> physical GPU (device: 0, name: Tesla P40, pci bus id: 0000:83:00.0, compute capability: 6.1)
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 0 into logs/model.ckpt.
INFO:tensorflow:loss = 0.7005657, step = 0
INFO:tensorflow:global_step/sec: 13.5045
INFO:tensorflow:loss = 0.3813903, step = 100 (7.405 sec)
INFO:tensorflow:global_step/sec: 15.0218
INFO:tensorflow:loss = 0.5762321, step = 200 (6.657 sec)
INFO:tensorflow:global_step/sec: 15.0521
INFO:tensorflow:loss = 0.15642221, step = 300 (6.644 sec)
INFO:tensorflow:global_step/sec: 15.0839
INFO:tensorflow:loss = 0.19744259, step = 400 (6.630 sec)
INFO:tensorflow:global_step/sec: 15.0664
INFO:tensorflow:loss = 0.110218376, step = 500 (6.638 sec)
INFO:tensorflow:global_step/sec: 15.0758
INFO:tensorflow:loss = 0.04312685, step = 600 (6.633 sec)
INFO:tensorflow:Saving checkpoints for 625 into logs/model.ckpt.
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2018-08-16-01:50:14
INFO:tensorflow:Graph was finalized.
2018-08-16 09:50:15.665224: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0
2018-08-16 09:50:15.665298: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-08-16 09:50:15.665314: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 
2018-08-16 09:50:15.665324: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N 
2018-08-16 09:50:15.665545: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 21544 MB memory) -> physical GPU (device: 0, name: Tesla P40, pci bus id: 0000:83:00.0, compute capability: 6.1)
INFO:tensorflow:Restoring parameters from logs/model.ckpt-625
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Evaluation [10/100]
INFO:tensorflow:Evaluation [20/100]
INFO:tensorflow:Evaluation [30/100]
INFO:tensorflow:Evaluation [40/100]
INFO:tensorflow:Evaluation [50/100]
INFO:tensorflow:Evaluation [60/100]
INFO:tensorflow:Evaluation [70/100]
INFO:tensorflow:Finished evaluation at 2018-08-16-01:50:19
INFO:tensorflow:Saving dict for global step 625: accuracy = 0.5, global_step = 625, loss = 7912.4326
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 625: logs/model.ckpt-625
INFO:tensorflow:Loss for final step: 0.028676154.
```
...

**Comment by rohan100jain** [comment]
One thing to look out for is making sure the shapes of the predictions and labels being fed into the accuracy metric are the same. There could be some broadcasting at times that can lead to odd values for the accuracy number. Use tf.shape. tf.Print to print out for a few steps and see whats going on.

**Comment by MustafaMustafa** [comment]
@honeytidy, did you resolve this issue? I am having a very similar issue. In fact, I replaced the eval dataset with the training dataset and I still get different training and eval losses on the same dataset.

**Comment by Zhiqiang-Ye** [comment]
@honeytidy did you solve your problem?

**Comment by wissebarkhof** [comment]
I experience a similar problem where the evaluation loss reported during training (using `estimator.train_and_evaluate()`)at a given step is much higher than when I evaluate the checkpoint at this step using `estimator.evaluate()`. The datasets and `EvalSpec` are exactly the same. I already changed the `steps` paramter in the `EvalSpec` to `None`. Is this behavior expected?

**Comment by mhajiaghayi** [comment]
I had a similar problem and it turned out to be related to the batch normalization module in my code that the fact that I needed to update the moving average and variance each time before training. basically adding the following commands would fix this issue. 

```
   optimizer = tf.train.AdamOptimizer(FLAGS.learning_rate)
    with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
        train_op = optimizer.minimize(loss_fn, var_list=tf.trainable_variables())
```
